src/services/ghost/classic-auto-complete/GhostContextProvider.ts

```diff
@@ -0,0 +1,89 @@
+import * as vscode from "vscode"
+import { ContextRetrievalService } from "../../continuedev/core/autocomplete/context/ContextRetrievalService"
+import { VsCodeIde } from "../../continuedev/core/vscode-test-harness/src/VSCodeIde"
+import { AutocompleteInput } from "../types"
+import { AutocompleteSnippetType } from "../../continuedev/core/autocomplete/snippets/types"
+import { HelperVars } from "../../continuedev/core/autocomplete/util/HelperVars"
+import { getAllSnippetsWithoutRace } from "../../continuedev/core/autocomplete/snippets/getAllSnippets"
+import { getDefinitionsFromLsp } from "../../continuedev/core/vscode-test-harness/src/autocomplete/lsp"
+import { DEFAULT_AUTOCOMPLETE_OPTS } from "../../continuedev/core/util/parameters"
+import { getSnippets } from "../../continuedev/core/autocomplete/templating/filtering"
+import { formatSnippets } from "../../continuedev/core/autocomplete/templating/formatting"
+
+function convertToContinuedevInput(autocompleteInput: AutocompleteInput) {
+	return {
+		...autocompleteInput,
+		recentlyVisitedRanges: autocompleteInput.recentlyVisitedRanges.map((range) => ({
+			...range,
+			type: AutocompleteSnippetType.Code,
+		})),
+	}
+}
+
+export class GhostContextProvider {
+	private contextService: ContextRetrievalService
+	private ide: VsCodeIde
+
+	constructor(context: vscode.ExtensionContext) {
+		this.ide = new VsCodeIde(context)
+		this.contextService = new ContextRetrievalService(this.ide)
+	}
+
+	/**
+	 * Get the IDE instance for use by tracking services
+	 */
+	public getIde(): VsCodeIde {
+		return this.ide
+	}
+
+	/**
+	 * Get context snippets for the current autocomplete request
+	 * Returns comment-based formatted context that can be added to prompts
+	 */
+	async getFormattedContext(autocompleteInput: AutocompleteInput, filepath: string): Promise<string> {
+		try {
+			// Convert filepath to URI if it's not already one
+			const filepathUri = filepath.startsWith("file://") ? filepath : vscode.Uri.file(filepath).toString()
+
+			// Initialize import definitions cache
+			await this.contextService.initializeForFile(filepathUri)
```

Comment:
this doesnt seem thread safe; we can have multiple requests in flight, and then we initialize it for one file but by the time we create the snippetPayload it might have already been changed to another file

Reply:
That is indeed an edgecase, but I still don't think that has any (serious) consequences.

Reply:
Ok, I looked into the implementation.  This isn't a concurrency bug, for sure; it's an optimization that's used in continue but not (currently) in our codebase, based on code comments.

Continue apparently _prefetched_ this data (some tree-sitter based parse tree) when the user switching editor tab, so that when an autocomplete request is made it's already in the cache. I don't see where we do, but I'll believe the comments that continue does do that.

Harmless; ignorable. Even if for some weird reason the parse was evicted due to some race condition, the code would simply return zero snippets as a result of that parse; not critical - and it also seems highly unlikely, anyhow.

Reply:
To save this knowledge for later: https://github.com/Kilo-Org/kilocode/pull/3506/files#r2510281102